src/llama-grammar.cpp

```diff
@@ -4,75 +4,530 @@
 #include "llama-sampling.h"
 
 #include <algorithm>
+#include <stdexcept>
 
-// Decodes a UTF-8 string which may end in an incomplete sequence. Adds a terminating 0 for use as
-// pointer. If an invalid sequence is encountered, returns `llama_partial_utf8.n_remain == -1`.
-std::pair<std::vector<uint32_t>, llama_partial_utf8> decode_utf8(
-        const std::string & src,
-        llama_partial_utf8 partial_start) {
-    static const int      lookup[] = { 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 2, 2, 3, 4 };
-    const char          * pos      = src.c_str();
-    std::vector<uint32_t> code_points;
+//
+// helpers
+//
 
-    // common english strings have the same number of codepoints and bytes. `+ 1` for the terminating 0.
-    code_points.reserve(src.size() + 1);
-    uint32_t value    = partial_start.value;
-    int      n_remain = partial_start.n_remain;
+// NOTE: assumes valid utf8 (but checks for overrun)
+// TODO: deduplicate
+static std::pair<uint32_t, const char *> decode_utf8(const char * src) {
+    static const int lookup[] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
```

Comment:
I already posted this message on python-llama-cpp repository as the developer recoded the whole thing in python.

Wouldn't this be an issue if the first byte value > 16 ? unless I'm missing something

For example, you can't use this kind of characters https://www.fileformat.info/info/unicode/char/1e958/index.htm

Reply:
Also it doesn't seem to follow communication bytes "directive", not sure it really matters here.

```cpp
static const uint8_t utf8_length_lookup[256] = {
    // 0x00-0x7F: 1 byte
    [0x00 ... 0x7F] = 1,
    // 0x80-0xBF: continuation bytes, should not be first byte
    [0x80 ... 0xBF] = 0,
    // 0xC0-0xDF: 2 bytes
    [0xC0 ... 0xDF] = 2,
    // 0xE0-0xEF: 3 bytes
    [0xE0 ... 0xEF] = 3,
    // 0xF0-0xF7: 4 bytes
    [0xF0 ... 0xF7] = 4,
    // 0xF8-0xFF: invalid for UTF-8
    [0xF8 ... 0xFF] = 0
};

uint8_t first_byte = static_cast<uint8_t>(*src);
int num_bytes = utf8_length_lookup[first_byte];
```

Reply:
nevermind